rubicon/objc/api.py

```diff
@@ -1554,42 +1552,47 @@ def __subclasscheck__(self, subclass):
 NSObjectProtocol = ObjCProtocol('NSObject')
 
 
-# Instances of DeallocationObserver are associated with every
-# ObjCInstance that needs to be kept alive, which is the case
-# when the ObjCInstance has had additional Python attributes added to it.
-# Their sole purpose is to watch for when the Objective-C object
-# is deallocated, and then remove the object from the
-# ObjCInstance._keep_alive_objects dictionary.
+# When a Python object is assigned to a new ObjCInstance attribute, the Python
+# object should be kept alive for the lifetime of the ObjCInstance. This is
+# done by wrapping the Python object as a WrappedPyObject that manages increments
+# the reference count during assignment and decrements it the the WrappedPyObject
+# and the owning ObjCInstance are deallocated.
 #
 # The methods of the class defined below are decorated with
-# rawmethod() instead of method() because DeallocationObservers
+# rawmethod() instead of method() because WrappedPyObject
 # are created inside of ObjCInstance's __new__ method and we have
 # to be careful to not create another ObjCInstance here (which
 # happens when the usual method decorator turns the self argument
 # into an ObjCInstance), or else get trapped in an infinite recursion.
 
-# Try to reuse an existing DeallocationObserver class.
+# Try to reuse an existing WrappedPyObject class.
 # This allows reloading the module without having to restart
-# the interpreter, although any changes to DeallocationObserver
+# the interpreter, although any changes to WrappedPyObject
 # itself are only applied after a restart of course.
 try:
-    DeallocationObserver = ObjCClass("DeallocationObserver")
+    WrappedPyObject = ObjCClass("WrappedPyObject")
 except NameError:
-    class DeallocationObserver(NSObject):
+    class WrappedPyObject(NSObject):
 
-        observed_object = objc_ivar(objc_id)
+        wrapped_pointer = objc_ivar(c_long)
 
         @objc_rawmethod
-        def initWithObject_(self, cmd, anObject):
+        def initWithObjectId_(self, cmd, address: c_long):
             self = send_message(self, 'init', restype=objc_id, argtypes=[])
             if self is not None:
-                set_ivar(self, 'observed_object', anObject)
+                pyo = cast(address, py_object)
+                pyo_ptr = cast(id(pyo.value), POINTER(PyObject))
+                pythonapi.Py_IncRef(pyo_ptr)
+                set_ivar(self, 'wrapped_pointer', c_long(address))
             return self.value
 
         @objc_rawmethod
         def dealloc(self, cmd) -> None:
-            anObject = get_ivar(self, 'observed_object')
-            ObjCInstance._keep_alive_objects.pop(anObject.value, None)
+            address = get_ivar(self, 'wrapped_pointer')
+            if address.value:
+                pyo = cast(address.value, py_object)
+                pyo_ptr = cast(id(pyo.value), POINTER(PyObject))
```

Comment:
Like above - is the `pyo` variable needed here?

```suggestion
                pyo_ptr = cast(address.value, POINTER(PyObject))
```